apps/sel4test-tests/src/arch/arm/tests/breakpoints.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2024, UNSW Sydney
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <autoconf.h>
+
+#ifdef CONFIG_HARDWARE_DEBUG_API
+
+#include <sel4/sel4.h>
+#include <utils/attribute.h>
+
+#include <arch/debug.h>
+#include "../../../helpers.h"
+#include "../../../test.h"
+#include "../../../tests/breakpoints.h"
+
+static int faulter_main(seL4_Word arg0, seL4_Word arg1, seL4_Word arg2, seL4_Word arg3)
+{
+    TEST_SINGLE_STEP_ASM();
+    /* NOTHING BELOW SHOULD EVER BE REACHED */
+    int *ptr = NULL;
+    *ptr = 0xBEEF;
+    return 0;
+}
+
+static int handler_main(seL4_Word faulter_tcb, seL4_Word a1, seL4_Word a2, seL4_Word a3)
+{
+    seL4_Word sender_badge;
+    seL4_MessageInfo_t tag;
+    seL4_Word label;
+
+    /* Wait for the faulter to trigger an event. */
+    tag = api_wait(fault_ep_cspath.capPtr, &sender_badge);
+    ZF_LOGV("Handler got a fault on the ep.\n");
+    label = seL4_MessageInfo_get_label(tag);
+
+    /* Check that it is a breakpoint exception */
+    if (label != seL4_Fault_DebugException ||
+        seL4_GetMR(seL4_DebugException_ExceptionReason) != seL4_InstructionBreakpoint) {
+        ZF_LOGV("Received exception other than instruction breakpoint exception.\n");
+        return -1;
+    }
+
+    /* Disable the breakpoint and enable single stepping */
+    int error = seL4_TCB_UnsetBreakpoint(faulter_tcb, 0);
+    test_eq(error, 0);
+
+    seL4_TCB_ConfigureSingleStepping_t res = seL4_TCB_ConfigureSingleStepping(faulter_tcb, 0, 1);
+    test_eq(res.error, 0);
+
+    /* Reply to the thread to resume it */
+    tag = api_reply_recv(fault_ep_cspath.capPtr, seL4_MessageInfo_new(0, 0, 0, 0), &sender_badge, 0);
+    label = seL4_MessageInfo_get_label(tag);
+
+    /* Wait for the single step exception */
+    fault_data.vaddr = seL4_GetMR(seL4_DebugException_FaultIP);
+    fault_data.reason = seL4_GetMR(seL4_DebugException_ExceptionReason);
+    fault_data.vaddr2 = seL4_GetMR(seL4_DebugException_TriggerAddress);
+    fault_data.bp_num = seL4_GetMR(seL4_DebugException_BreakpointNumber);
+
+    if (label == seL4_Fault_DebugException &&
+        fault_data.reason == seL4_SingleStep &&
+        fault_data.vaddr == (seL4_Word) &TEST_SS_LABEL_TWO) {
```

Comment:
You could add this check for 'TEST_SS_LABEL_ONE' above too. If the test passes it doesn't matter, but if something went wrong with the initial breakpoint it would make debugging a lot easier.